# FINAL PROJECT

In [73]:
# The code was removed by Watson Studio for sharing.

In [74]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
%matplotlib inline

In [75]:
data=df_data_1
data.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y

In [76]:
data.shape

(614, 13)

In [77]:
data['Loan_Status'].replace(('Y','N'),(1,0),inplace=True)
data['Loan_Status'].value_counts()

1    422
0    192
Name: Loan_Status, dtype: int64

In [78]:
df=data.copy()
df['Self_Employed'].value_counts()

No     500
Yes     82
Name: Self_Employed, dtype: int64

In [79]:
df.head()

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area  Loan_Status  
0             1.0         Urban            1  
1             1.0         Rural            0  
2             1.0         Urban            1  
3             1.0         Urban            1  
4             1.0         Urban            1

In [80]:
df['Gender'].replace(('Male','Female'),(1,0),inplace=True)
df['Married'].replace(('Yes','No'),(1,0),inplace=True)
df['Education'].replace(('Graduate','Not Graduate'),(1,0),inplace=True)
df.head()

Loan_ID  Gender  Married Dependents  Education Self_Employed  \
0  LP001002     1.0      0.0          0          1            No   
1  LP001003     1.0      1.0          1          1            No   
2  LP001005     1.0      1.0          0          1           Yes   
3  LP001006     1.0      1.0          0          0            No   
4  LP001008     1.0      0.0          0          1            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   

   Credit_History Property_Area  Loan_Status  
0             1.0         Urban            1  
1             1.0         Rural            0  
2             1.0         Urban            1  
3             1.0         Urban            1  
4             1.0         Urban            1

In [81]:
cdf=df.drop(['Loan_ID','Dependents','Property_Area'],axis=1)
cdf.head()

Gender  Married  Education Self_Employed  ApplicantIncome  \
0     1.0      0.0          1            No             5849   
1     1.0      1.0          1            No             4583   
2     1.0      1.0          1           Yes             3000   
3     1.0      1.0          0            No             2583   
4     1.0      0.0          1            No             6000   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                0.0         NaN             360.0             1.0   
1             1508.0       128.0             360.0             1.0   
2                0.0        66.0             360.0             1.0   
3             2358.0       120.0             360.0             1.0   
4                0.0       141.0             360.0             1.0   

   Loan_Status  
0            1  
1            0  
2            1  
3            1  
4            1

In [82]:
cdf.fillna(0,inplace=True)
cdf['Self_Employed'].replace(('Yes','No'),(1,0),inplace=True)
cdf.head()

Gender  Married  Education  Self_Employed  ApplicantIncome  \
0     1.0      0.0          1              0             5849   
1     1.0      1.0          1              0             4583   
2     1.0      1.0          1              1             3000   
3     1.0      1.0          0              0             2583   
4     1.0      0.0          1              0             6000   

   CoapplicantIncome  LoanAmount  Loan_Amount_Term  Credit_History  \
0                0.0         0.0             360.0             1.0   
1             1508.0       128.0             360.0             1.0   
2                0.0        66.0             360.0             1.0   
3             2358.0       120.0             360.0             1.0   
4                0.0       141.0             360.0             1.0   

   Loan_Status  
0            1  
1            0  
2            1  
3            1  
4            1

In [83]:
cdf.dtypes

Gender               float64
Married              float64
Education              int64
Self_Employed          int64
ApplicantIncome        int64
CoapplicantIncome    float64
LoanAmount           float64
Loan_Amount_Term     float64
Credit_History       float64
Loan_Status            int64
dtype: object

# Correlating feature labels

In [84]:
cdf.corr()

Gender   Married  Education  Self_Employed  \
Gender             1.000000  0.339026  -0.061497      -0.027421   
Married            0.339026  1.000000  -0.017671       0.008491   
Education         -0.061497 -0.017671   1.000000       0.010383   
Self_Employed     -0.027421  0.008491   0.010383       1.000000   
ApplicantIncome    0.013933  0.050194   0.140760       0.127180   
CoapplicantIncome  0.082316  0.080496   0.062290      -0.016100   
LoanAmount         0.051540  0.140023   0.173244       0.109154   
Loan_Amount_Term  -0.052445 -0.079926   0.108857      -0.029185   
Credit_History     0.035780 -0.023626   0.081637      -0.004994   
Loan_Status        0.025407  0.084281   0.085884      -0.003700   

                   ApplicantIncome  CoapplicantIncome  LoanAmount  \
Gender                    0.013933           0.082316    0.051540   
Married                   0.050194           0.080496    0.140023   
Education                 0.140760           0.062290    0.173244   
Self_Employed             0.127180          -0.016100    0.109154   
ApplicantIncome           1.000000          -0.116605    0.538290   
CoapplicantIncome        -0.116605           1.000000    0.190377   
LoanAmount                0.538290           0.190377    1.000000   
Loan_Amount_Term         -0.015645          -0.050321    0.058519   
Credit_History            0.006986          -0.058795   -0.032711   
Loan_Status              -0.004710          -0.059187   -0.010631   

                   Loan_Amount_Term  Credit_History  Loan_Status  
Gender                    -0.052445        0.035780     0.025407  
Married                   -0.079926       -0.023626     0.084281  
Education                  0.108857        0.081637     0.085884  
Self_Employed             -0.029185       -0.004994    -0.003700  
ApplicantIncome           -0.015645        0.006986    -0.004710  
CoapplicantIncome         -0.050321       -0.058795    -0.059187  
LoanAmount                 0.058519       -0.032711    -0.010631  
Loan_Amount_Term           1.000000        0.050145     0.007306  
Credit_History             0.050145        1.000000     0.432616  
Loan_Status                0.007306        0.432616     1.000000

In [85]:
x=cdf[['Gender','Married','Education','Loan_Amount_Term','Credit_History']]
y=cdf['Loan_Status']

# Normalization of the data

In [86]:
#Normalization of data
from sklearn.preprocessing import StandardScaler
X=StandardScaler().fit_transform(x)
X

/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/preprocessing/data.py:645: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/base.py:464: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


array([[ 0.50559245, -1.3574213 ,  0.52836225,  0.31416232,  0.54095432],
       [ 0.50559245,  0.73669096,  0.52836225,  0.31416232,  0.54095432],
       [ 0.50559245,  0.73669096,  0.52836225,  0.31416232,  0.54095432],
       ...,
       [ 0.50559245,  0.73669096,  0.52836225,  0.31416232,  0.54095432],
       [ 0.50559245,  0.73669096,  0.52836225,  0.31416232,  0.54095432],
       [-1.97787765, -1.3574213 ,  0.52836225,  0.31416232, -1.84858491]])

# Splitting the data to Train and Test

In [87]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=4)
print("Train Set:",X_train.shape,y_train.shape)
print("Test Set:",X_test.shape,y_test.shape)

Train Set: (491, 5) (491,)
Test Set: (123, 5) (123,)


# Logistic Regression Model

In [88]:
from sklearn.linear_model import LogisticRegression
lreg=LogisticRegression(C=0.01,solver="liblinear").fit(X_train,y_train)
lreg

LogisticRegression(C=0.01, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='liblinear',
          tol=0.0001, verbose=0, warm_start=False)

In [89]:
pred=lreg.predict(X_test)
pred

array([1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,
       1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1])

In [90]:
print("Accuracy Score of Logistic Regression Model:",accuracy_score(y_test,lreg.predict(X_test))*100)

Accuracy Score of Logistic Regression Model: 78.86178861788618


# KNN Model

In [91]:
from sklearn.neighbors import KNeighborsClassifier

for k in range(3,50):
    reg=KNeighborsClassifier(n_neighbors=k).fit(X_train,y_train)
    acc_score=accuracy_score(y_test,reg.predict(X_test))
    if k==3:
        opt_k=k
        max_score=acc_score
        kreg=reg
    if acc_score>max_score:
        opt_k=k
        max_score=acc_score
        kreg=reg
print("K Neighbors Classifier:",kreg)        
print("Optimum K value is:",opt_k)
print("Accuracy Score:",max_score*100)

K Neighbors Classifier: KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=18, p=2,
           weights='uniform')
Optimum K value is: 18
Accuracy Score: 78.86178861788618


# Decision Tree Classifier Model

In [92]:
from sklearn.tree import DecisionTreeClassifier
dreg=DecisionTreeClassifier(criterion="entropy").fit(X_train,y_train)
dreg

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [93]:
dpred=dreg.predict(X_test)
dpred

array([1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0,
       1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1])

In [94]:
print("Decision Tree Model accuracy:",accuracy_score(y_test,dreg.predict(X_test))*100)

Decision Tree Model accuracy: 75.60975609756098


# SVM Model

In [95]:
from sklearn import svm
sreg = svm.SVC(kernel='rbf')
sreg.fit(X_train, y_train) 

/opt/conda/envs/Python36/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='rbf', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

In [96]:
spred=sreg.predict(X_test)
spred

array([1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0,
       0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0,
       1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1])

In [97]:
print("SVM Model Accuracy:",accuracy_score(y_test,spred)*100)

SVM Model Accuracy: 78.86178861788618


In [98]:
from sklearn.metrics import f1_score,log_loss

# F1 Scores

In [99]:
print("F1 Scores for TRAIN SET".center(110))
print("For Logistic Regression Model:",f1_score(y_train,lreg.predict(X_train)))
print("For KNN Model:",f1_score(y_train,kreg.predict(X_train)))
print("For Decision Tree Model:",f1_score(y_train,dreg.predict(X_train)))
print("For SVM Model:",f1_score(y_train,sreg.predict(X_train)))

                                           F1 Scores for TRAIN SET                                            
For Logistic Regression Model: 0.8396094839609484
For KNN Model: 0.8402203856749312
For Decision Tree Model: 0.848137535816619
For SVM Model: 0.8419580419580419


In [100]:
print("F1 Scores for TEST SET".center(110))
print("For Logistic Regression Model:",f1_score(y_test,lreg.predict(X_test)))
print("For KNN Model:",f1_score(y_test,kreg.predict(X_test)))
print("For Decision Tree Model:",f1_score(y_test,dreg.predict(X_test)))
print("For SVM Model:",f1_score(y_test,sreg.predict(X_test)))

                                            F1 Scores for TEST SET                                            
For Logistic Regression Model: 0.8555555555555555
For KNN Model: 0.8555555555555555
For Decision Tree Model: 0.8255813953488371
For SVM Model: 0.8539325842696628


# Log Loss

In [101]:
y_train_prob = lreg.predict_proba(X_train)
print("Log Loss for TRAIN SET".center(110))
print("Log Loss Value:",log_loss(y_train,y_train_prob))

                                            Log Loss for TRAIN SET                                            
Log Loss Value: 0.5653047110439261


In [102]:
y_test_prob = lreg.predict_proba(X_test)
print("Log Loss for TEST SET".center(110))
print("Log Loss Value:",log_loss(y_test,y_test_prob))

                                            Log Loss for TEST SET                                             
Log Loss Value: 0.5697163217254974


# Jaccard Similarity Score

In [103]:
from sklearn.metrics import jaccard_similarity_score

In [104]:
print("Jaccard Scores for TRAIN SET".center(110))
print("For Logistic Regression:",jaccard_similarity_score(y_train,lreg.predict(X_train))*100)
print("For KNN Model:",jaccard_similarity_score(y_train,kreg.predict(X_train))*100)
print("For Decision Tree Model:",jaccard_similarity_score(y_train,dreg.predict(X_train))*100)
print("For SVM Model:",jaccard_similarity_score(y_train,sreg.predict(X_train))*100)

                                         Jaccard Scores for TRAIN SET                                         
For Logistic Regression: 76.57841140529531
For KNN Model: 76.37474541751527
For Decision Tree Model: 78.41140529531569
For SVM Model: 76.9857433808554


In [105]:
print("Jaccard Scores for TEST SET".center(110))
print("For Logistic Regression:",jaccard_similarity_score(y_test,lreg.predict(X_test)))
print("For KNN Model:",jaccard_similarity_score(y_test,kreg.predict(X_test)))
print("For Decision Tree Model:",jaccard_similarity_score(y_test,dreg.predict(X_test)))
print("For SVM Model:",jaccard_similarity_score(y_test,sreg.predict(X_test)))

                                         Jaccard Scores for TEST SET                                          
For Logistic Regression: 0.7886178861788617
For KNN Model: 0.7886178861788617
For Decision Tree Model: 0.7560975609756098
For SVM Model: 0.7886178861788617


# Final Report

| __Algorithm__ | __F1-score__ | __Jaccard__ | __LogLoss__|
| --- | --- | --- | --- |
| Logistic Regression Model| 0.8555 | 0.7888 | 0.5697 |
| --- | --- | --- | --- |
| KNN Model| 0.8555 | 0.7888 | NA | 
| --- | --- | --- | --- |
| Decision Tree Model| 0.8208 | 0.7479 | NA |
| --- | --- | --- | --- |
| SVM Model| 0.8539 | 0.7886 | NA |

__Even though Decision Tree Model showed good result in training set prediction, but Logistic Regression Model and KNN Model out-performs other models in the test set prediction.
Logistic Regression Model is Best Model.__